keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/ReshapeVertex.java

```diff
@@ -0,0 +1,53 @@
+package io.improbable.keanu.vertices.dbl.nonprobabilistic.operators.unary;
+
+import io.improbable.keanu.tensor.TensorShape;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.vertices.Vertex;
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.diff.DualNumber;
+
+import java.util.Map;
+
+public class ReshapeVertex extends DoubleUnaryOpVertex {
+
+    private int[] proposedShape;
+
+    public ReshapeVertex(DoubleVertex inputVertex, int... proposedShape) {
+        super(inputVertex.getShape(), inputVertex);
+        this.proposedShape = proposedShape;
+    }
+
+    /**
+     * Returns the supplied vertex with a new shape of the same length
+     */
+    @Override
+    protected DoubleTensor op(DoubleTensor a) {
+        return a.reshape(proposedShape);
+    }
+
+    @Override
+    protected DualNumber calculateDualNumber(Map<Vertex, DualNumber> dualNumbers) {
+        Map<Long, DoubleTensor> partialDerivatives = dualNumbers.get(inputVertex).getPartialDerivatives().asMap();
+        int[] vertexShape = inputVertex.getShape();
+
+        for (Map.Entry<Long, DoubleTensor> partialDerivative : partialDerivatives.entrySet()) {
+            int[] shape = partialDerivative.getValue().getShape();
+            int[] wrtShape = extractWrtShape(shape, vertexShape);
+            int[] newPartialShape = TensorShape.concat(proposedShape, wrtShape);
+
+            DoubleTensor reshapedPartialDerivative = partialDerivative.getValue().reshape(newPartialShape);
+            partialDerivative.setValue(reshapedPartialDerivative);
```

Comment:
Make sure you don't mutate the partialDerivative of the parent. We should look into making the PartialDerivative class immutable.

Reply:
Oops - I was doing a ReshapeInPlaceVertex by mistake...

Changed to not mutate the parent